### PR TITLE
perf[ndjson]: use simd_json

### PR DESCRIPTION
**Pull request by universalmind303**
simd_json is [quite a bit faster](https://github.com/jorgecarleitao/json-deserializer/issues/11) for deserializing objects, it also exposes a lower level api where we can eventually do further optimizations by not deserializing unused fields.

The biggest drain on performance is the deserialization of unused data. 

for example 
```json
{"foo": 1, "bar": ["a", "b"]}
{"foo": 2, "bar": ["b", "c"]}
```
once we know the structure of the JSON, we don't need to deserialize `foo` into a string, we should be able to skip the it & only deserialize the values. Similarly, if the projection only contained `bar`, we should not deserialize `foo` or it's respective value. 

However, this is also valid JSON.
```json
{"bar": ["a", "b"], "foo": 1}
{"foo": 2, "bar": ["b", "c"]}
```
so if we implemented the reader in a way to not deserialize the keys, we would likely have to either infer that the data is ordered properly, or have an additional parameter for ordered/unordered json.

### CODE REVIEW

File: polars/polars-io/src/ndjson_core/ndjson.rs [hunk 1]
```diff
@@ -261,26 +252,26 @@ impl<'a> CoreJsonReader<'a> {
...
-    for value in stream.by_ref() {
-        let v = value.unwrap_or(Value::Null);
-        match v {
-            Value::Object(value) => {
-                buffers
-                    .iter_mut()
-                    .for_each(|(s, inner)| match value.get(s) {
+fn parse_lines<'a>(bytes: &[u8], buffers: &mut PlIndexMap<String, Buffer<'a>>) -> Result<()> {
+    for line in SplitLines::new(bytes, b'\n') {
+        let mut line = line.to_vec();
```
Comment (ritchie46) [inline polars/polars-io/src/ndjson_core/ndjson.rs:257]: Why do we need to own the line here?

**Comment by ritchie46** [comment]
Can you do a benchmark on a release build and some real world data? I like to know how much we improve.

> so if we implemented the reader in a way to not deserialize the keys, we would likely have to either infer that the data is ordered properly, or have an additional parameter for ordered/unordered json.

That would be a cool optimization indeed. Would be great if we can follow up on that.

File: polars/polars-io/src/ndjson_core/ndjson.rs (hunk 1 above)
Comment (universalmind303) [reply polars/polars-io/src/ndjson_core/ndjson.rs:257]: the docs on simd_json aren't great but from what I can tell, it looks like you always need an owned slice. 
```rust
// simd_json::to_borrowed_value
pub fn to_value(s: &mut [u8]) -> Result<Value>
```

Comment (universalmind303) [reply polars/polars-io/src/ndjson_core/ndjson.rs:257]: https://github.com/simd-lite/simd-json/issues/242

Comment (ritchie46) [reply polars/polars-io/src/ndjson_core/ndjson.rs]: > i think i could do a mem transmute to get around the ownership, but I'm not super comfortable with the unsafe apis.

If the parser mutates the slice, we are not allowed to do that.

We should at least keep an owned `Vec<u8>` buffer around. That way we only pay for the `memcpy` and not for the allocation.

Comment (universalmind303) [reply polars/polars-io/src/ndjson_core/ndjson.rs]: > We should at least keep an owned Vec<u8> buffer around. 
are there any examples of this in the repo?

Comment (ritchie46) [reply polars/polars-io/src/ndjson_core/ndjson.rs]: In pseudo rust code you do this:

```rust
fn parse_impl<'a>(bytes: &[u8], buffers: &mut PlIndexMap<String, Buffer<'a>>, line: &mut Vec<u8>) -> Result<usize> {
    // here we reuse the buffer every iteration
    line.clear();
    line.extend_from_slice(bytes);

    match line.len() {
        0 => Ok(0),
        1 => {
            if line[0] == NEWLINE {
                Ok(1)
            } else {
                Err(PolarsError::ComputeError(
                    "Invalid JSON: unexpected end of file".into(),
                ))
            }

// other code
...
  let mut buf = vec![];
  for line in SplitLines::new(bytes, NEWLINE) {
        offset += 1; // the newline
        offset += parse_impl(line, buffers, &mut buf)?;
    }

// other code

```

File: polars/polars-io/src/json.rs [hunk 2]
```diff
@@ -242,7 +242,7 @@ where
 
 #[cfg(test)]
 mod test {
-    use std::io::Cursor;
+    use std::{ffi::OsStr, io::Cursor};
```
Comment (ritchie46) [inline polars/polars-io/src/json.rs:245]: Can we move the tests to the integration tests? I want to have most test there as it is easier on compile times.

**Comment by universalmind303** [comment]
@ritchie46 some benchmark data on a release build

Tested on a [~700mb JSON file](https://drive.google.com/file/d/1hKgcJcr8paz8SqMXGKViZl8L6KTIioPV/view?usp=sharing
), ran 20 iterations of `pl.scanJson('my_file.json').collectSync()`
## nodejs-polars 0.6
```js
// durations are in milliseconds
mean: 3170, 
min: 3130, 
max: 3190, 
std: 18.237,
```

## simd_json
```js
mean: 310, 
min: 248, 
max: 352, 
std: 39.153
```

a mean of about 10x faster.  🚀 🚀

**Comment by jorgecarleitao** [comment]
Impressive!! 🚀

**Comment by ritchie46** [comment]
Wow.. that's a lot! 💪

**Comment by universalmind303** [comment]
since we always know the keys we are looking for, it makes sense to use the `known-key` feature in simd-json. Which gives us a pretty significant performance gain. 

with `known-key` object access. 🚀 🚀 
```js
mean: 251, 
min: 236, 
max: 263, 
std: 8.083 
```

**Comment by ritchie46** [comment]
> 

Nice. That was the optimization we were talking about. Just a feature flag away 🙂

**Comment by Licenser** [comment]
<3 This is wonderful to see :)

**Pull request merged**